File: 0079-word-search/0079-word-search.java
class Solution {
    private static int m, n;
    private static final int[] dx = {0,1,0,-1}, dy= {1,0,-1,0};
    public boolean exist(char[][] board, String word) {
        m = board.length;
        n = board[0].length;
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                    if(exist(i, j, word, 0, board)) return true;
            }
        }
        return false;
    }
    private boolean exist(int r, int c, String targetWord, int idx, char[][] board){
        
        if(idx>=targetWord.length()) 
            return true;

        if(r< 0 || r>=m || c<0 || c>=n || board[r][c]!=targetWord.charAt(idx))
            return false;

        boolean result;
        board[r][c] = '#';
        for(int i=0; i<4; i++){
            if(exist(r+dx[i], c+dy[i], targetWord, idx+1, board)) return true;
        }
        board[r][c] = targetWord.charAt(idx);
        return false;
    } 
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        m, n = len(board), len(board[0])
        dx = [0,1,0,-1]
        dy = [1,0,-1,0]
        def word_searched(row, col, idx):
            if idx == len(word):
                return True
            
            if row < 0 or row >= m or col < 0 or col >= n or board[row][col] != word[idx]:
                return False
            
            tmp = board[row][col]
            board[row][col] = '#'
            for i in range(4):
                n_row = row + dx[i]
                n_col = col + dy[i]
                if word_searched(n_row, n_col, idx+1):
                    return True
            
            board[row][col] = tmp
            return False
        
        for i in range(m):
            for j in range(n):
                if word_searched(i,j,0):
                    return True
        
        return False
        
        